source/common/protobuf/utility.cc

```diff
@@ -556,17 +560,57 @@ void redact(Protobuf::Message* message, bool ancestor_is_sensitive) {
       // `MessageUtil::redact()` for more context.
       ENVOY_LOG_MISC(warn, "Could not redact udpa::type::v1::TypedStruct with unknown type URL {}",
                      typed_struct->type_url());
-      return;
+      return false;
     }
 
     MessageUtil::jsonConvert(typed_struct->value(), ProtobufMessage::getNullValidationVisitor(),
                              *typed_message);
     redact(typed_message.get(), ancestor_is_sensitive);
     MessageUtil::jsonConvert(*typed_message, *(typed_struct->mutable_value()));
+    return true;
+  }
+  return false;
+}
+
+// `DataSource`, which is annotated as `sensitive` in `TlsCertificate` and `TlsSessionTicketKeys`
+// requires special handling.
+template <typename DataSource>
+bool redactDataSource(Protobuf::Message* message, bool ancestor_is_sensitive) {
+  auto* data_source = dynamic_cast<DataSource*>(message);
+  if (data_source != nullptr) {
+    if (ancestor_is_sensitive) {
+      switch (data_source->specifier_case()) {
+      case DataSource::SPECIFIER_NOT_SET:
+        // If the data source is empty, no work is needed.
+        break;
+      case DataSource::kFilename:
+        // Don't redact filenames (SecretManagerImplTest::ConfigDumpNotRedactFilenamePrivateKey).
+        break;
+      case DataSource::kInlineBytes:
+        // Clear inline bytes and treat it as a string (fall through).
+        data_source->clear_inline_bytes();
+      case DataSource::kInlineString:
+        // Redact strings the usual way.
+        data_source->set_inline_string("[redacted]");
+        break;
+      }
+    }
+    return true;
+  }
+  return false;
+}
+
+// Recursive helper method for MessageUtil::redact() below. Note that we have to keep track of
+// whether an ancestor was marked as `sensitive`, not just the field, to handle cases similar to
+// `DataSource` in a more general way.
+void redact(Protobuf::Message* message, bool ancestor_is_sensitive) {
+  if (redactAny(message, ancestor_is_sensitive) ||
+      redactTypedStruct(message, ancestor_is_sensitive) ||
+      redactDataSource<envoy::api::v2::core::DataSource>(message, ancestor_is_sensitive) ||
+      redactDataSource<envoy::api::v3alpha::core::DataSource>(message, ancestor_is_sensitive)) {
```

Comment:
You shouldn't need to manually write this for v3alpha (or use a template above). Envoy will be automatically upgraded to v3alpha internally in the coming week or two, so please stick to v2 for now.

Reply:
Got it, thanks